GeneratorInterface/Core/interface/ConcurrentHadronizerFilter.h

```diff
@@ -119,6 +130,12 @@ namespace edm {
     unsigned int counterRunInfoProducts_;
     unsigned int nAttempts_;
     mutable std::atomic<gen::StreamCache<HAD, DEC>*> useInLumi_{nullptr};
+    mutable std::atomic<unsigned long long> greatestNStreamEndLumis_{0};
+    mutable std::atomic<bool> streamEndRunComplete_{true};
+    // The next two data members are thread safe and can be safely mutable because
+    // they are only modified/read in globalBeginRun and globalBeginLuminosityBlock.
```

Comment:
Can we (in principle) have globalBeginRun and globalBeginLuminosityBlock transitions being run simultaneously? (obviously the lumi would have to belong to a different run) Or two globalBeginRuns or two globalBeginLuminosityBlocks?

Reply:
This is an interesting question. I would be interested in hearing Chris's (@Dr15Jones) opinion about it.

The existing code in CMSSW serializes globalBeginLumi transitions (and all run transitions). The globalBeginLumi transition completes before the Framework even looks to see what transition should come next. The run concurrency PR currently under review continues that pattern for globalBeginRun. globalBeginLumi and globalBeginRun are serialized like input source activity. They will not run concurrently. It would not be trivial to change this. It would take some effort just to identify the things that would break if we changed this behavior. I think it is built into the current Framework implementation in multiple places.

On the other hand, the TWIKI that documents the multithreading design does not require this serialization: https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkTransitions
Maybe we should add this requirement to this TWIKI documentation?

If we want to make this code support that kind of concurrency in addition to adding support for run concurrency then I need to do more work on this PR (and in multiple other places probably in separate PRs later).

My personal guess is that the performance improvement this additional concurrency would allow is not worth the effort that would be required to implement it.

Reply:
The global run/lumi transitions need to be serialized as the EventSetup system doesn't support concurrent determination of new IOVs.

Reply:
Thanks. I think we should document that the global run/lumi transitions are serialized (I didn't read the twiki David pointed to carefully enough to say whether it would be implied by the text there already or not)

Reply:
I added an extra requirement in the TWIKI with this additional requirement on global begin transitions (in the bullet points under the figure, the next to last one is new). Feel free to edit it or let me know if you want me to reword it.

https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkTransitions

Reply:
Thanks David, looks good.